tests/topotests/multicast-pim-sm-topo2/test_multicast_pim_sm_topo2.py

```diff
@@ -0,0 +1,1953 @@
+#
+# Permission to use, copy, modify, and/or distribute this software
+# for any purpose with or without fee is hereby granted, provided
+# that the above copyright notice and this permission notice appear
+# in all copies.
+#
+# THE SOFTWARE IS PROVIDED "AS IS" AND VMWARE DISCLAIMS ALL WARRANTIES
+# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL VMWARE BE LIABLE FOR
+# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY
+# DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
+# WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
+# ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
+# OF THIS SOFTWARE.
+#
+
+"""
+Following tests are covered to test multicast pim sm:
+
+Test steps
+- Create topology (setup module)
+- Bring up topology
+
+Following tests are covered:
+1. TC_17: Verify (*,G) and (S,G) present and multicast traffic resume,
+    after restart of PIMd daemon
+2. TC_18: Verify (*,G) and (S,G) present and multicast traffic resume after
+    FRR service stop and start
+3. TC_10: Verify SPT switchover working when RPT and SPT path is
+    different
+4. TC_15: Verify (S,G) and (*,G) mroute after shut / no shut of upstream
+    interfaces
+5. TC_7: Verify mroute detail when receiver is present
+    outside of FRR
+6. TC_8: Verify mroute when FRR is acting as FHR and LHR
+7. TC_20: Verify mroute detail when 5 different receiver joining
+    same source
+8. TC_22: Verify OIL and IIF detail updated in (S,G) mroute after shut
+    and no shut of the source interface
+"""
+
+import os
+import sys
+import json
+import time
+import datetime
+from time import sleep
+import pytest
+
+# Save the Current Working Directory to find configuration files.
+CWD = os.path.dirname(os.path.realpath(__file__))
+sys.path.append(os.path.join(CWD, "../"))
+sys.path.append(os.path.join(CWD, "../lib/"))
+
+# Required to instantiate the topology builder class.
+
+# pylint: disable=C0413
+# Import topogen and topotest helpers
+from lib.topogen import Topogen, get_topogen
+from mininet.topo import Topo
+
+from lib.common_config import (
+    start_topology,
+    write_test_header,
+    write_test_footer,
+    step,
+    iperfSendIGMPJoin,
+    addKernelRoute,
+    reset_config_on_routers,
+    iperfSendTraffic,
+    kill_iperf,
+    shutdown_bringup_interface,
+    kill_router_daemons,
+    start_router,
+    start_router_daemons,
+    stop_router,
+    required_linux_kernel_version,
+    topo_daemons,
+)
+from lib.pim import (
+    create_pim_config,
+    create_igmp_config,
+    verify_igmp_groups,
+    verify_ip_mroutes,
+    verify_pim_interface_traffic,
+    verify_upstream_iif,
+    verify_pim_neighbors,
+    verify_pim_state,
+    verify_ip_pim_join,
+    clear_ip_mroute,
+    clear_ip_pim_interface_traffic,
+    verify_igmp_config,
+)
+from lib.topolog import logger
+from lib.topojson import build_topo_from_json, build_config_from_json
+
```

Comment:
Please add pytest module level markers as described in dev guide: http://docs.frrouting.org/projects/dev-guide/en/latest/topotests.html#markers

Reply:
Done